src/server/stream_family.cc

```diff
@@ -1040,55 +1042,144 @@ void XGroupHelp(CmdArgList args, ConnectionContext* cntx) {
   return (*cntx)->SendSimpleStrArr(help_arr);
 }
 
-}  // namespace
+OpResult<int64_t> OpTrim(const OpArgs& op_args, const AddTrimOpts& opts) {
+  auto* shard = op_args.shard;
+  auto& db_slice = shard->db_slice();
+  OpResult<PrimeIterator> res_it = db_slice.Find(op_args.db_cntx, opts.key, OBJ_STREAM);
+  if (!res_it) {
+    if (res_it.status() == OpStatus::KEY_NOTFOUND) {
+      return 0;
+    }
+    return res_it.status();
+  }
 
-void StreamFamily::XAdd(CmdArgList args, ConnectionContext* cntx) {
-  string_view key = ArgS(args, 0);
-  unsigned id_indx = 1;
-  AddOpts add_opts;
+  CompactObj& cobj = (*res_it)->second;
+  stream* s = (stream*)cobj.RObjPtr();
+
+  int64_t deleted = 0;
+  if (!opts.limit) {
+    if (opts.trim_strategy == TrimStrategy::kMaxLen) {
+      /* Notify xtrim event if needed. */
+      deleted = streamTrimByLength(s, opts.max_len, opts.trim_approx);
+      // TODO: when replicating, we should propagate it as exact limit in case of trimming.
+    } else if (opts.trim_strategy == TrimStrategy::kMinId) {
+      deleted = streamTrimByID(s, opts.minid.val, opts.trim_approx);
+    }
+  } else {
+    streamAddTrimArgs trim_args = {};
+    trim_args.trim_strategy = static_cast<int>(opts.trim_strategy);
+    trim_args.approx_trim = opts.trim_approx;
+    trim_args.limit = opts.limit;
+
+    if (opts.trim_strategy == TrimStrategy::kMaxLen) {
+      trim_args.maxlen = opts.max_len;
+    } else if (opts.trim_strategy == TrimStrategy::kMinId) {
+      trim_args.minid = opts.minid.val;
+    }
+    deleted = streamTrim(s, &trim_args);
+  }
+
```

Comment:
Isn't it the same as in OpAdd? I think we can unite it in a common trim function

Reply:
yep sure - updated